lgc/patch/RegisterMetadataBuilder.cpp

```diff
@@ -197,34 +197,33 @@ void RegisterMetadataBuilder::buildEsGsRegisters() {
   const auto &calcFactor = gsInOutUsage.gs.calcFactor;
   const auto tesResUsage = m_pipelineState->getShaderResourceUsage(ShaderStageTessEval);
   const auto &tesBuiltInUsage = tesResUsage->builtInUsage.tes;
+  const bool hasTs = m_hasTcs || m_hasTes;
 
   // ES_VGPR_COMP_CNT in SPI_SHADER_PGM_RSRC2_GS
   unsigned gsVgprCompCnt = 0;
   if (calcFactor.inputVertices > 4 || gsBuiltInUsage.invocationId)
-    gsVgprCompCnt = 3;
+    gsVgprCompCnt = 3; // Enable vtx4/vtx5 offset (GS VGPR3) or GS instance ID (GS VGPR4)
   else if (gsBuiltInUsage.primitiveIdIn)
-    gsVgprCompCnt = 2;
+    gsVgprCompCnt = 2; // Enable primitive ID (GS VGPR2)
   else if (calcFactor.inputVertices > 2)
-    gsVgprCompCnt = 1;
+    gsVgprCompCnt = 1; // Enable vtx2/vtx3 offset (GS VGPR1)
   getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::GsVgprCompCnt] = gsVgprCompCnt;
 
   // ES_VGPR_COMP_CNT in SPI_SHADER_PGM_RSRC2_GS
   unsigned esVgprCompCnt = 0;
-  if (m_hasTcs || m_hasTes) {
-    // NOTE: when primitive ID is used, set vgtCompCnt to 3 directly because primitive ID is the last VGPR.
+  if (hasTs) {
     if (tesBuiltInUsage.primitiveId)
-      esVgprCompCnt = 3;
+      esVgprCompCnt = 3; // Enable patch ID (ES VGPR8)
     else
-      esVgprCompCnt = 2;
-
-    if (m_pipelineState->isTessOffChip())
-      getHwShaderNode(Util::Abi::HardwareStage::Gs)[Util::Abi::HardwareStageMetadataKey::OffchipLdsEn] = true;
+      esVgprCompCnt = 2; // Must enable relative patch ID (ES VGPR7)
   } else {
     if (vsBuiltInUsage.instanceIndex)
-      esVgprCompCnt = 3; // Enable instance ID
+      esVgprCompCnt = 3; // Enable instance ID (ES VGPR8)
   }
   getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::EsVgprCompCnt] = esVgprCompCnt;
 
+  getHwShaderNode(Util::Abi::HardwareStage::Gs)[Util::Abi::HardwareStageMetadataKey::OffchipLdsEn] = hasTs;
```

Comment:
Since `OffchipLdsEn` is `HardwareStage::Gs`-specific, it's better to set it here. Could you help remove the duplicate setting of it in `buildShaderExecutionRegisters` L988.

Reply:
OK. Fixed.